app/com/gu/floodgate/reindex/BulkJobActor.scala

```diff
@@ -0,0 +1,219 @@
+    "Packages Editor",
+    "Section Manager",
+    "Tag Manager",
+    "Flexible Content"
+  )
+  private val orderedContentSources = sortContentToReindex
+  private val nameIdMap = orderedContentSources.map(source => (source.id, source.appName)).toMap
+  private var pendingReindexes: List[(String, String)] = Nil // our queue
+  private val PollInterval = 200.milliseconds
+
+  final def receive: Receive = notReindexing
+
+  private def notReindexing: Receive = {
+    case StartBulkReindex => {
+      if (noRunningReindexes) {
+        pendingReindexes = orderedContentSources.map(contentSource => (contentSource.id, contentSource.environment))
+        sender ! CanTrigger
+        become(reindexing)
+        self ! ReindexNext
+      } else {
+        sender ! RejectJob
+      }
+    }
+    case IsActorReindexing => sender ! NotReindexing
+    case other => logger.warn(s"Unexpected message received: $other")
+  }
+
+  private def reindexing: Receive = {
+    case ReindexNext => reindexNextContentSource
+    case ReindexStarted(id, env) => {
+      pendingReindexes = pendingReindexes.filterNot(contentTuple => contentTuple == (id, env))
+      self ! Polling(id, env)
+    }
+    case ReindexError(id, env, error) => {
+      pendingReindexes = Nil
+      val failedReindex = contentSources.find(contentSource => (contentSource.id, contentSource.environment) == (id, env))
+      logger.warn(s"Unable to trigger Reindex for $failedReindex")
+      become(notReindexing)
+    }
+    case Polling(id, env) => pollRunningReindexes(id, env)
+    case IsActorReindexing => buildReindexInfo pipeTo sender()
+    case DropPendingJob(id, env) => {
+      pendingReindexes = pendingReindexes.filterNot(contentTuple => (id, env) == contentTuple)
+      if (pendingReindexes.contains((id, env)) || runningJobService.getRunningJob(id, env).isRight) {
+        sender ! FailedToCancelPendingJob
+      } else {
+        sender ! CancelledPendingJob
+      }
+    }
+  }
+
+  private def buildReindexInfo: Future[IsReindexing] = {
+
+    val futureRunningJobs = getRunningJobs
+    val futureCompletedJobs = getCompletedJobs
+
+    for {
+      runningJobs <- futureRunningJobs
+      completedJobs <- futureCompletedJobs
+      pendingJobs = getPendingJobs
+    } yield {
+      IsReindexing(runningJobs, pendingJobs, completedJobs)
+    }
+  }
+
+  private def getRunningJobs: Future[Seq[RunningJobInfo]] = {
+    val env = orderedContentSources.map(source => source.environment).distinct
+    val futureRunningJobs = runningJobService.getAllRunningJobs().map(jobs => jobs.filter(job => env.contains(job.contentSourceEnvironment)))
+    futureRunningJobs.map { runningJobs =>
+      runningJobs.map { job =>
+        val name = nameIdMap.getOrElse(job.contentSourceId, s"Name not found for ID: ${job.contentSourceId}")
+        val contentSource = orderedContentSources.find(contentSource => (contentSource.id, contentSource.environment) == (job.contentSourceId, job.contentSourceEnvironment))
```

Comment:
there are a number of places where we have to scan `pendingReindexes` or `orderedContentSources`. I wonder if it would be simpler to make `pendingReindexes` a `List[ContentSource]`? It might make the code simpler